substrate/primitives/consensus/beefy/src/lib.rs

```diff
@@ -352,6 +387,171 @@ where
 	return valid_first && valid_second
 }
 
+/// Checks whether the provided header's payload differs from the commitment's payload.
+fn check_header_proof<Header>(
+	commitment: &Commitment<Header::Number>,
+	canonical_header: &Option<Header>,
+	canonical_header_hash: &Header::Hash,
```

Comment:
Please clarify - I'm probably missing something.

> Looking at full path for building and later validating

"full path for building and later validating" suggests to me you're looking at the client side? The `canonical_header_hash != canonical_header.hash()` would indeed be redundant on a client with an archival database (or for a recent block) since they could just retrieve the header from their own database.

But for the proof verification inside the runtime, the `canonical_header_hash` is *the* anchor of truth since the runtime retrieves this hash from its own state; everything else non-trivial is supplied by the prover.
Without this check, the prover could fabricate a header whose digest's mmr root at `commitment.block_number` differs from the commitment's (which the prover claims to equivocate), and could slash honest validators with this fabrication.

The only alternative I can see for avoiding this check is the prover also providing a justification for the header to show it's been finalized. But this is both logically and computationally more involved than just checking its hash against the known source of truth.